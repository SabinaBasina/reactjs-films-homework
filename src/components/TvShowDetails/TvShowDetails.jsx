import React, { Component } from 'react';
import http from 'axios';
import styles from './TvShowDetails.scss';
import Loading from '../TvShowPage/Loading.gif';
import NoImage from '../TvShow/NoImage.jpg';
import Episodes from '../Episodes/Episodes';

class TvShowDetails extends Component {
  constructor(props) {
    super(props);

    this.state = {
      tvShow: {},
      isReady: false,
    };

    // eslint-disable-next-line react/prop-types
    const { id } = this.props;

    http
      .get(`https://api.tvmaze.com/shows/${id}`)
      .then((response) => {
        this.setState({ tvShow: response.data, isReady: true });
      });
  }


  render() {
    const { isReady, tvShow } = this.state;
    return (
      <div>

        {!isReady
           && (
           <img
             src={Loading}
             className={styles.Ready}
             alt=""
           />
           )}

        {isReady

            && (
            <div className={styles.TvShow}>

              <h1>{tvShow.name}</h1>

              <div className={styles.Information}>

                <img
                  src={tvShow.image === undefined ? NoImage : tvShow.image.medium}
                  className={styles.TvShowImage}
                  alt=""
                />

                <div>
                  <p dangerouslySetInnerHTML={{ __html: tvShow.summary }} />
                  <br />
                  <b>Genres:</b>
                  {' '}
                  {tvShow.genres && tvShow.genres.join(', ')}
                  <br />
                  <b>Country:</b>
                  {' '}
                  {tvShow.network && tvShow.network.country.name}
                  <br />
                  <b>Premiered:</b>
                  {' '}
                  <time>{tvShow.premiered}</time>
                </div>

              </div>


              <p className={styles.Episodes}>Episodes</p>
              <Episodes nameTvShow={tvShow.name} />

            </div>
            )
        }

      </div>
    );
  }
}

export default TvShowDetails;
