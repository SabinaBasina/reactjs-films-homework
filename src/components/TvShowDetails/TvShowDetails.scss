.Ready {
  display: block;
  margin: 30px auto 0 auto;
  text-align: center;
}

.TvShow {
  text-align: justify;
}

.TvShow h1 {
  text-align: center;
  margin-top: 40px;
  margin-bottom: 20px;
  font-weight: 500;
}

.Information {
  display: block;
  margin-left: 20%;
  margin-right: 20%;
}

.Information::after {
  content: " ";
  display: table;
  clear: both;
}

.TvShowImage {
  float: left;
  margin: 0 35px 15px 0;
}

.Episodes {
  margin: 100px 15% 60px 15%;
  height: 35px;
  background-color: #222;
  border-radius: 8px;
  color: white;
  font-weight: 100;
  font-size: 24px;
  text-align: center;
}

@media only screen and (max-width: 1100px) {
  .Information {
    margin-left: 10%;
    margin-right: 10%;
  }

  .Episodes {
    margin-left: 5%;
    margin-right: 5%;
  }
}

@media only screen and (max-width: 820px) {
  .Episodes {
    margin-left: 0;
    margin-right: 0;
  }

  .TvShowImage {
    display: block;
    margin: 0 auto 40px auto;
    float: none;
  }

  .Information {
    margin-left: 5%;
    margin-right: 5%;
  }
}
