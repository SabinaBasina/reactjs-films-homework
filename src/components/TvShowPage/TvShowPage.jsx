import React, { Component } from 'react';
// import http from 'axios';
import PropTypes from 'prop-types';
import TvShow from '../TvShow/TvShow';
import Search from '../Search';
import styles from './TvShowPage.scss';
import Loading from './Loading.gif';

class TvShowPage extends Component {
  state = {
    isReady: false,
  };


  componentDidMount() {
    const { loadTvShows } = this.props;
    loadTvShows();
    this.setReadyState(true);
  }

  componentDidUpdate(prevProps) {
    const { page, loadTvShows } = this.props;
    const { isReady } = this.state;
    if ((page !== prevProps.page) && isReady) {
      loadTvShows();
    }
  }

  setReadyState = (newIsReady) => {
    const { onReady } = this.props;
    this.setState({ isReady: newIsReady });
    onReady(newIsReady);
  }

  render() {
    // const { isReady } = this.state;
    const { tvShows } = this.props;

    if (!tvShows) {
      return null;
    }
    return (

      <main className={styles.TvShows}>

        {!tvShows
          && (
            <img
              src={Loading}
              className={styles.Ready}
              alt="Loading page"
            />
          )}

        {tvShows
          && (
            <div>
              <Search />
              <div className={styles.TvShowLibrary}>
                {tvShows.map(tvShowData => (
                  <TvShow data={tvShowData} />
                ))}
              </div>
            </div>
          )}

      </main>

    );
  }
}

export default TvShowPage;

TvShowPage.propTypes = {
  loadTvShows: PropTypes.func,
  onReady: PropTypes.func,
  page: PropTypes.number,
  tvShows: PropTypes.arrayOf(PropTypes.array).isRequired,
};

TvShowPage.defaultProps = {
  loadTvShows: () => { },
  onReady: () => { },
  page: 0,
};
