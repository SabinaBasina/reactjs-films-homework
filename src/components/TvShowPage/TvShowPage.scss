.TvShows {
  text-align: center;
}

.TvShowLibrary {
  margin-left: 15%;
  margin-right: 15%;
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(260px, 1fr));
}
