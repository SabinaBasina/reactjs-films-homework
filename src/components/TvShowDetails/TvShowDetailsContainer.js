import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import TvShowDetails from './TvShowDetails';
import { loadTvShowsDetails } from '../../redux/actions/tvshowdetails';


const mapStateToProps = createStructuredSelector({
  tvShow: state => state.tvshowdetails.tvShow,
});

const mapDispatchToProps = {
  loadTvShowsDetails,
};

export default connect(mapStateToProps, mapDispatchToProps)(TvShowDetails);
