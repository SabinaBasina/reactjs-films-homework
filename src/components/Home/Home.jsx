import React, { Component } from 'react';
import PropTypes from 'prop-types';

import styles from './Home.scss';
import TvShowPage from '../TvShowPage';
import Loading from '../../images/Loading.gif';

class Home extends Component {
  increment = () => {
    const { onIncrement } = this.props;
    onIncrement();
  }

  decrement = () => {
    const { onDecrement } = this.props;
    onDecrement();
  }

  render() {
    const { page, tvShows } = this.props;
    return (

      <main>

        <TvShowPage page={page} />

        {!tvShows
          && (
            <img
              src={Loading}
              className={styles.Ready}
              alt="Loading page"
            />
          )}

        {tvShows
          && (
            <div className={styles.Pager}>

              {page > 0
                && (
                  <button type="button" onClick={e => this.decrement(e)}>
                    &larr; Previous Page
                  </button>
                )
              }

              <button type="button" onClick={e => this.increment(e)}>
                Next Page &rarr;
              </button>

            </div>
          )}


      </main>

    );
  }
}

export default Home;

Home.propTypes = {
  page: PropTypes.number,
  tvShows: PropTypes.instanceOf(Object),
  onIncrement: PropTypes.func,
  onDecrement: PropTypes.func,
};

Home.defaultProps = {
  page: 0,
  tvShows: undefined,
  onIncrement: () => { },
  onDecrement: () => { },
};
