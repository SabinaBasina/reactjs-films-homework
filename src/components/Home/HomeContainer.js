import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import Home from './Home';

const mapStateToProps = createStructuredSelector({
  page: state => state.page.page,
});

const mapDispatchToProps = {
  onIncrement: () => ({ type: 'INCREMENT' }),
  onDecrement: () => ({ type: 'DECREMENT' }),
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
