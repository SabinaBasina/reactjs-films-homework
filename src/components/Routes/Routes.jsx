/* eslint-disable react/prefer-stateless-function */
import React, { PureComponent, Fragment } from 'react';
import { Route } from 'react-router-dom';

import './Routes.scss';
import Home from '../../pages/Home';
import TvShowDetails from '../../pages/TvShowDetails';
import TvShowPage from '../TvShowPage';

class Routes extends PureComponent {
  render() {
    return (
      <Fragment>
        <Route path="/" exact component={Home} />
        <Route
          path="/page/:pageNumber"
          render={props => <Home pageNumber={Number(props.match.params.pageNumber)} />}
        />
        <Route
          path="/search/:searchValue"
          render={props => <TvShowPage searchValue={props.match.params.searchValue} />}
        />
        <Route
          path="/tvShowDetails/:id"
          render={props => <TvShowDetails id={props.match.params.id} />}
        />
      </Fragment>
    );
  }
}

export default Routes;
