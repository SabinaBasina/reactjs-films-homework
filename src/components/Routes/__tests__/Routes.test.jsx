import React from 'react';
import TestRenderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router';

import Routes from '../Routes';
import TvShowDetails from '../../../pages/TvShowDetails';
import Home from '../../../pages/Home';
import TvShowPage from '../../TvShowPage';

jest.mock('../../../pages/TvShowDetails', () => jest.fn(() => 'TvShowDetails'));
jest.mock('../../TvShowPage', () => jest.fn(() => 'TvShowPage'));
jest.mock('../../../pages/Home', () => jest.fn(() => 'Home'));


describe('1', () => {
  test('renders 1', () => {
    const component = TestRenderer.create(
      <MemoryRouter initialEntries={['/tvShowDetails/1']} initialIndex={0}>
        <Routes />
      </MemoryRouter>,
    );
    expect(component).toMatchSnapshot();
    expect(TvShowDetails.mock.calls[0][0]).toEqual({ id: '1' });
  });

  test('renders 2', () => {
    const component = TestRenderer.create(
      <MemoryRouter initialEntries={['/search/a']} initialIndex={0}>
        <Routes />
      </MemoryRouter>,
    );
    expect(component).toMatchSnapshot();
    expect(TvShowPage.mock.calls[0][0]).toEqual({ searchValue: 'a' });
  });

  test('renders 3', () => {
    const component = TestRenderer.create(
      <MemoryRouter initialEntries={['/page/1']} initialIndex={0}>
        <Routes />
      </MemoryRouter>,
    );
    expect(component).toMatchSnapshot();
    expect(Home.mock.calls[0][0]).toEqual({ pageNumber: '1' });
  });
});
