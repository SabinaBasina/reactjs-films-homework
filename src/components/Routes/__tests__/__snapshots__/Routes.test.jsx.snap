// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 renders 1 1`] = `"TvShowDetails"`;

exports[`1 renders 2 1`] = `"TvShowPage"`;

exports[`1 renders 3 1`] = `"Home"`;
